pkg/gatewayroute/resource_manager.go

```diff
@@ -0,0 +1,357 @@
+package gatewayroute
+
+import (
+	"context"
+	appmesh "github.com/aws/aws-app-mesh-controller-for-k8s/apis/appmesh/v1beta2"
+	"github.com/aws/aws-app-mesh-controller-for-k8s/pkg/aws/services"
+	"github.com/aws/aws-app-mesh-controller-for-k8s/pkg/conversions"
+	"github.com/aws/aws-app-mesh-controller-for-k8s/pkg/k8s"
+	"github.com/aws/aws-app-mesh-controller-for-k8s/pkg/mesh"
+	"github.com/aws/aws-app-mesh-controller-for-k8s/pkg/references"
+	"github.com/aws/aws-app-mesh-controller-for-k8s/pkg/runtime"
+	"github.com/aws/aws-app-mesh-controller-for-k8s/pkg/virtualgateway"
+	"github.com/aws/aws-app-mesh-controller-for-k8s/pkg/virtualservice"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	appmeshsdk "github.com/aws/aws-sdk-go/service/appmesh"
+	"github.com/go-logr/logr"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/conversion"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ResourceManager is dedicated to manage AppMesh GatewayRoute resources for k8s GatewayRoute CRs.
+type ResourceManager interface {
+	// Reconcile will create/update AppMesh GatewayRoute to match gr.spec, and update gr.status
+	Reconcile(ctx context.Context, gr *appmesh.GatewayRoute) error
+
+	// Cleanup will delete AppMesh GatewayRoute created for gr.
+	Cleanup(ctx context.Context, gr *appmesh.GatewayRoute) error
+}
+
+func NewDefaultResourceManager(
+	k8sClient client.Client,
+	appMeshSDK services.AppMesh,
+	referencesResolver references.Resolver,
+	accountID string,
+	log logr.Logger) ResourceManager {
+
+	return &defaultResourceManager{
+		k8sClient:          k8sClient,
+		appMeshSDK:         appMeshSDK,
+		referencesResolver: referencesResolver,
+		accountID:          accountID,
+		log:                log,
+	}
+}
+
+// defaultResourceManager implements ResourceManager
+type defaultResourceManager struct {
+	k8sClient          client.Client
+	appMeshSDK         services.AppMesh
+	referencesResolver references.Resolver
+	accountID          string
+	log                logr.Logger
+}
+
+func (m *defaultResourceManager) Reconcile(ctx context.Context, gr *appmesh.GatewayRoute) error {
+	ms, err := m.findMeshDependency(ctx, gr)
```

Comment:
Is this necessary? wouldn't the below find + validate virtualGatewayDependency validate the Mesh as well? VirtualGateway wouldn't be up and active if the corresponding mesh is missing right?